Rewrite lesson and tests

gj opened the pull request:
Fixes #4 
Fixes #6

gj committed:
Rewrite confusing sections in README

gj committed:
Update package.json

gj committed:
Add placeholder comments
to show where to copy over code from the README

gj committed:
Fix test suite
so that it actually tests something beyond that the functions are defined
(#4 #6)

jmburges commented on README.md:
````diff
@@ -24,95 +22,135 @@ Well, it's our lucky day!
+```
 
-For example, let's say that we have a function for logging the square of a number called, appropriately enough, `square`.
+...but that seems like too many extra lines for such a simple operation. We could also use an IIFE, an immediately-invoked function expression, to make our code a bit more compact, but it's starting to get far more confusing than the elegant simplicity of a `.forEach()`:
````
Drop the bit about IIFE. They are going to do research and just generally go down the rabbit hole that is IIFEs. Either we have to do a whole thing on it..or not explain it. Remember this is early curriculum for some

jmburges commented on README.md:
```diff
@@ -24,95 +22,135 @@ Well, it's our lucky day!
+`.forEach()` provides a wonderfully compact, readable syntax for invoking a function on every member of an array. It's a powerful tool that makes it easier for us to perform complex array operations in just a few lines of code.
+
+## Functions as Arguments
+Functions in JavaScript are **first-class objects**, and one of the best perks of that special status is that we can pass functions as arguments to other functions. Oftentimes, these passed-in functions serve as **callbacks**, which do just what the name implies: they are _called back_, potentially with arguments, and do their work on data that weren't available at runtime.
```
What does First-class objects mean? Effectively it means that functions can be treated as data just like variables are. What's the difference between a string and a function in terms of just hunks of data floating around? NOTHING. That's what first-class objects mean

jmburges replied:
Who are they called back by? Nobody ever understands who is actually doing the calling back

jmburges replied:
Also I think we go into callbacks more later. Tell the students that because it's another rabbit hole they'll fall down

jmburges commented on test/index-test.js:
```diff
@@ -1,45 +1,49 @@
-    it('completely alters an array in place when used with Array.prototype.forEach', () => {
-      const array = [1, 2, 3]
+    it('invokes the passed-in callback function on every element of the passed-in array using Array.prototype.forEach()', () => {
+      const callback = fruit => `Mmmm, ${fruit}!!!`;
```
Do they know aray functions at this stage? Even if they do...they just *barely* know it. I would stick to regular functions here. It's important for students to understand that a callback is just a function passed in as an argument

jmburges replied:
(I know it's in the Test...we want students to read and understand our tests)

gj replied:
I hear you, but the lesson is on `Array.prototype.forEach()`, which is explicitly for operating on an array. The preceding lesson is all about callbacks. Or am I completely missing the point you're making? 😅

jmburges replied:
I can't type. I meant *arrow* functions.

gj replied:
Should I also avoid arrow functions for the `describe()` and `it()` callbacks?

gj committed:
Remove note on IIFEs
to not confuse students or send them down an unnecessary rabbit hole

gj committed:
Replace arrow function with regular function expression
to not confuse students with arrow functions at this early point in the JS curriculum

gj committed:
Add explanation of callbacks in JavaScript
